File: AllInOne/src/Array/MergeIntervals.java
package src.Array;

import java.util.List;
import java.util.LinkedList;
import java.util.Arrays;
/**
 * 
 * 56. Merge Intervals
 * 
 * @author jingjiejiang
 * @history Apr 26, 2022
 * 
 */
public class MergeIntervals {
  public int[][] merge(int[][] intervals) {
        
    assert intervals != null && intervals.length > 1;

    List<int[]> resList = new LinkedList<>();

    Arrays.sort(intervals, (intval1, intval2) -> intval1[0] != intval2[0] ?
                            intval1[0] - intval2[0] : intval1[1] - intval2[1]);
    
    int[] preIntval = intervals[0]; 

    for (int row = 1; row < intervals.length; row ++) {

      if (intervals[row][0] <= preIntval[1]) {
        preIntval[1] = Math.max(intervals[row][1], preIntval[1]);
      } else { // post[0] > pre[1]
        resList.add(preIntval);
        preIntval = intervals[row];
      }
    }

    resList.add(preIntval);

    int[][] resArr = new int[resList.size()][2];

    return resList.toArray(resArr);
  }
}

File: AllInOne/src/Array/MergeIntervals.js
/**
 * 
 * 56. Merge Intervals
 * 
 */
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
 var merge = function(intervals) {
    
  if (!intervals.length) return [];
  intervals.sort((a, b) => a[0] === b[0] ? a[1] - b[1] : a[0] - b[0]);

  const res = [intervals[0]];
  const prev = intervals[0];

  for (let [start, end] of intervals) {
    if (start <= prev.end) {
      prev.end = Math.max(prev.end, end);
    } else {
      res.push([start, end]);
      prev = [start, end];
    }
  }

  return res;
};